packages/devextreme/testing/tests/DevExpress.ui.widgets/tabs.tests.js

```diff
@@ -422,6 +423,200 @@ QUnit.module('General', () => {
+            assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+        });
+
+        QUnit.test('Selection should not change when onSelectionChanging is cancelled with a rejected promise', function(assert) {
+            const onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = new Promise((resolve, reject) => {
+                    setTimeout(() => {
+                        reject('Cancellation error');
+                    }, this.CANCEL_DELAY);
+                });
+            });
+
+            this.tabInstance.option('onSelectionChanging', onSelectionChangingSpy);
+            this.tabInstance.option('onSelectionChanged', this.onSelectionChangedSpy);
+            const $item = this.$tabs.find(`.${TABS_ITEM_CLASS}`).eq(1);
+
+            $item.trigger('dxclick');
+
+            assert.strictEqual(onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should be called twice after click');
+            assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should not be called before promise resolves');
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 before promise resolves');
+            assert.deepEqual(this.tabInstance.option('selectedItem'), { text: '0' }, 'selectedItem should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItems'), [{ text: '0' }], 'selectedItems should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+
+            this.clock.tick(this.CANCEL_DELAY);
+
+            onSelectionChangingSpy.getCall(0).args[0].cancel.catch((error) => {
+                assert.strictEqual(error, 'Cancellation error', 'Promise should be rejected with "Cancellation error"');
+                assert.strictEqual(onSelectionChangingSpy.callCount, 1, 'onSelectionChanging should still be called once');
+                assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should be called after promise rejection');
+                assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should change after cancellation promise rejection');
+                assert.deepEqual(this.tabInstance.option('selectedItem'), { text: '0' }, 'selectedItem should remain the first item');
+                assert.deepEqual(this.tabInstance.option('selectedItems'), [{ text: '0' }], 'selectedItems should remain the first item');
+                assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+            });
+        });
+
+        QUnit.test('Selection should be cancelled asynchronously when e.cancel is resolved', function(assert) {
```

Comment:
[test coverage]
Let's add a similar tests checking that selection is proceed correctly when promise is resolved with false